app/src/main/java/com/google/jetpackcamera/MainActivity.kt

```diff
@@ -101,13 +107,67 @@ class MainActivity : ComponentActivity() {
                             modifier = Modifier.fillMaxSize(),
                             color = MaterialTheme.colorScheme.background
                         ) {
-                            JcaApp(onPreviewViewModel = { previewViewModel = it })
+                            JcaApp(
+                                onPreviewViewModel = { previewViewModel = it },
+                                contentResolver = contentResolver,
+                                imageCaptureUri = uri,
+                                onImageCapture = { event ->
+                                    val shouldFinishWithResult = isExternalImageCapture()
+                                    when (event) {
+                                        is CameraUseCase.ImageCaptureEvent.ImageSaved -> {
+                                            if (shouldFinishWithResult) {
+                                                setResult(RESULT_OK)
+                                                finish()
+                                            }
+                                        }
+
+                                        is CameraUseCase.ImageCaptureEvent.ImageCaptureError -> {
+                                            if (shouldFinishWithResult) {
+                                                setResult(RESULT_CANCELED)
+                                                finish()
+                                            }
+                                        }
+                                    }
+                                }
+                            )
```

Comment:
woulnd't this violate the  MediaStore contract? If we use this mode class, then it would also encapsulate the uri that's supposed to be at EXTRA_OUTOUT or clipData?

Reply:
Ah I see what you are saying. The mode is generated in mainActivity from the extra (or lack of) then passed into JCAApp. Yeah we could certainly do it that way

Reply:
It shouldn't change anything about how we fulfill the contract, it's just a way to organize the code without having to add new arguments to the `JcaApp` constructor for every intent `ACTION_*` that we handle.

You could even just encapsulate the arguments that you already have in the `ExternalImageCaptureMode` subclass:
```
sealed interface JcaAppMode

class StandardMode() : JcaAppMode

data class ExternalImageCaptureMode(
    val contentResolver: ContentResolver,
    val imageCaptureUri: Uri?,
    val onImageCapture: (ImageCaptureEvent) -> Unit
) : JcaAppMode
```

Then the constructor for JcaApp is:
```
class JcaApp(
    onPreviewViewModel = { previewViewModel = it },
    val appMode: JcaAppMode = StandardMode
)
```

Then for the `captureImage` call, you can do:
```
when (appMode) {
  is StandardMode -> captureImage() // previous "captureImage" overload with no args
  is ExternalImageCaptureMode ->
      captureImage(appMode.contentResolver, appMode.imageCaptureUri, appMode.onImageCapture)
}
```

Eventually we can use the `JcaAppMode` to determine how the UI looks. For instance, in `ExternalImageCapture` mode, there is no need to allow users to capture video, so we may show a special icon for the capture button that makes it more clear that video is disabled.